File: kafka/kafka-annotation-processor/src/main/java/ru/tinkoff/kora/kafka/annotation/processor/consumer/KafkaConsumerModuleGenerator.java
package ru.tinkoff.kora.kafka.annotation.processor.consumer;

import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.TypeSpec;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.kafka.annotation.processor.KafkaClassNames;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;

public class KafkaConsumerModuleGenerator {
    private final Elements elements;
    private final KafkaConsumerHandlerGenerator kafkaConsumerHandlerGenerator;
    private final KafkaConsumerConfigGenerator configGenerator;
    private final KafkaConsumerContainerGenerator kafkaConsumerContainerGenerator;

    public KafkaConsumerModuleGenerator(ProcessingEnvironment processingEnv, KafkaConsumerHandlerGenerator kafkaConsumerHandlerGenerator, KafkaConsumerConfigGenerator configGenerator, KafkaConsumerContainerGenerator kafkaConsumerContainerGenerator) {
        this.elements = processingEnv.getElementUtils();
        this.kafkaConsumerHandlerGenerator = kafkaConsumerHandlerGenerator;
        this.configGenerator = configGenerator;
        this.kafkaConsumerContainerGenerator = kafkaConsumerContainerGenerator;
    }

    public final JavaFile generateModule(TypeElement typeElement) {
        var classBuilder = TypeSpec.interfaceBuilder(typeElement.getSimpleName().toString() + "Module")
            .addOriginatingElement(typeElement)
            .addAnnotation(AnnotationUtils.generated(KafkaConsumerModuleGenerator.class))
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(CommonClassNames.module);

        for (var element : typeElement.getEnclosedElements()) {
            if (element.getKind() != ElementKind.METHOD) {
                continue;
            }
            var method = (ExecutableElement) element;
            var annotation = AnnotationUtils.findAnnotation(method, KafkaClassNames.kafkaListener);
            if (annotation == null) {
                continue;
            }
            var configTagData = this.configGenerator.generate(elements, method, annotation);
            classBuilder.addMethod(configTagData.configMethod());
            if (configTagData.tag() != null) {
                classBuilder.addType(configTagData.tag());
            }

            var parameters = ConsumerParameter.parseParameters(method);
            var handler = this.kafkaConsumerHandlerGenerator.generate(elements, method, parameters);
            classBuilder.addMethod(handler.method());

            var container = this.kafkaConsumerContainerGenerator.generate(elements, method, annotation, handler, parameters);
            classBuilder.addMethod(container);
        }

        var packageName = this.elements.getPackageOf(typeElement);
        return JavaFile.builder(packageName.getQualifiedName().toString(), classBuilder.build()).build();
    }
}

File: kafka/kafka-symbol-processor/src/main/kotlin/io/koraframework/kafka/symbol/processor/consumer/KafkaConsumerModuleGenerator.kt
package io.koraframework.kafka.symbol.processor.consumer

import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.TypeSpec
import io.koraframework.kafka.symbol.processor.KafkaClassNames
import io.koraframework.ksp.common.AnnotationUtils.findAnnotation
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.KspCommonUtils.addOriginatingKSFile
import io.koraframework.ksp.common.KspCommonUtils.generated

class KafkaConsumerModuleGenerator(
    private val kafkaHandlerGenerator: KafkaHandlerGenerator,
    private val kafkaConfigGenerator: KafkaConsumerConfigGenerator,
    private val kafkaContainerGenerator: KafkaContainerGenerator
) {
    fun generateModule(declaration: KSClassDeclaration): FileSpec {
        val classBuilder = TypeSpec.interfaceBuilder(declaration.simpleName.asString() + "Module")
            .addOriginatingKSFile(declaration)
            .generated(KafkaConsumerModuleGenerator::class)

        classBuilder.addAnnotation(AnnotationSpec.builder(CommonClassNames.module).build())
        for (function in declaration.getDeclaredFunctions()) {
            val kafkaListener = function.findAnnotation(KafkaClassNames.kafkaListener)
            if (kafkaListener == null) {
                continue
            }

            val configTagData = kafkaConfigGenerator.generate(function, kafkaListener)
            classBuilder.addFunction(configTagData.configFunction)
            if (configTagData.tag != null) {
                classBuilder.addType(configTagData.tag)
            }

            val parameters = ConsumerParameter.parseParameters(function)

            val handler = kafkaHandlerGenerator.generate(function, parameters)
            classBuilder.addFunction(handler.funSpec)

            val container = kafkaContainerGenerator.generate(function, kafkaListener, handler, parameters)
            classBuilder.addFunction(container)
        }
        val packageName = declaration.packageName.asString()
        return FileSpec.get(packageName, classBuilder.build())
    }
}
